arch/risc-v/include/arch.h

```diff
@@ -120,6 +245,84 @@ extern "C"
 #define EXTERN extern
 #endif
 
+#ifndef __ASSEMBLY__
+
+/****************************************************************************
```

Comment:
We cannot have APIs with riscv_ prefix in public arch.h, do we need these APIs ?

Reply:
How about change prefix to `arch_`

Reply:
public arch api should start with up_

Reply:
OK

Reply:
but we only need move riscv_percpu_s to public heder files, all other definition could still keep in riscv_percpu.h

Reply:
`up_percpu_set_thread` is needed here

Reply:
I think we should do it so that tcb is forced to be the first element of CSR_SCRATCH, and we can access tcb by simply reading the scratch register's first index. This way we don't need to expose the public API and we get the maximum speed benefit.

Reply:
Agree, so let's force tcb to be the firest element

Reply:
It doesn't work even at the first.

Reply:
There is a extra memory read operation by this approach compared to ARM's, but it works:
```c
#define up_this_task() (*((struct tcb_s **)READ_CSR(CSR_SCRATCH)))
```

Reply:
Why riscv works, but arm not? both need at least two instructions to fetch task pointer. As far as I know, only x64 arch support read task pointer from cpu specific storage with one instruction.

Reply:
`READ_CSR(CSR_SCRATCH)` will read `CSR_SCRATCH` to a common register on RISCV, then it can be used to access memorys.

Reply:
@xiaoxiang781216 is right, accessing CSR_SCRATCH->xxx is only safe when a context switch (CPU switch) cannot occur, meaning local interrupts have to be disabled.

We need to use TP for the information, but it does not work for FLAT mode I think?

Reply:
For this issue, these reading logics also should be protected by irq save/ irq restore:
https://github.com/apache/nuttx/blob/ffdbbdeef368bf31f067cc8d8941dda35f4ee84b/arch/risc-v/src/common/riscv_percpu.c#L172-L201

Reply:
Then again, holding the return value of either outside of a critical section is not valid either ?

Reply:
@pussuw TP is used for thread local storage if it's enabled

Reply:
Yes I know, in protected/kernel mode we could use TP still, but in flat mode it's impossible :(

Reply:
@pussuw If these API are not called frequently, how about store the TCB pointer in scratch register, and use another approach to get percpu_s , such as fetch it from g_percpu by hart id?

Reply:
They are accessed via assembly macros, not via C API.

Reply:
For example the stack pointers are needed immediately when entering exception_common. You cannot replace this functionality with anything else. I'm sorry, you need to think of something else.

Reply:
How about add a new option to control this behavior?

Allow the user to choose between using percpu registers to store the TCB to accelerate scheduling or maintaining the current functionality.

Reply:
I guess I don't have any objections to this. It can work in flat mode.

Btw you could use tp in protected / kernel mode (CONFIG_SYSCALL=y), just swap the user tp with CSR_SCRATCH->tcb in exception_common, and update tp when a context switch occurs. This way even the protected builds could enjoy the speed benefit?

Reply:
why not always enable the acceleration to simplify the code base, I don't see other arch provide this type of flexibility.

Reply:
OK, let's implement it for flat build first